Library/Homebrew/cask/migrator.rb

```diff
@@ -0,0 +1,86 @@
+# typed: true
+# frozen_string_literal: true
+
+require "cask/cask_loader"
+require "utils/inreplace"
+
+module Cask
+  class Migrator
+    extend ::Utils::Inreplace
+
+    attr_reader :old_cask, :new_cask
+
+    sig { params(old_cask: Cask, new_cask: Cask).void }
+    def initialize(old_cask, new_cask)
+      raise CaskNotInstalledError, new_cask unless new_cask.installed?
+
+      @old_cask = old_cask
+      @new_cask = new_cask
+    end
+
+    sig { params(new_cask: Cask, dry_run: T::Boolean).void }
+    def self.migrate_if_needed(new_cask, dry_run: false)
+      old_tokens = new_cask.old_tokens
+      return if old_tokens.empty?
+
+      return unless (installed_caskfile = new_cask.installed_caskfile)
+
+      installed_token = installed_caskfile.relative_path_from(Caskroom.path).basename.to_s
+      return if new_cask.token == installed_token
+
+      old_cask = CaskLoader.load(installed_caskfile)
+      migrator = new(old_cask, new_cask)
+      migrator.migrate(dry_run: dry_run)
+    end
+
+    sig { params(dry_run: T::Boolean).void }
+    def migrate(dry_run: false)
+      old_token = old_cask.token
+      new_token = new_cask.token
+
+      old_caskroom_path = old_cask.caskroom_path
+      new_caskroom_path = new_cask.caskroom_path
+
+      old_installed_caskfile = old_cask.installed_caskfile.relative_path_from(old_caskroom_path)
+      new_installed_caskfile = old_installed_caskfile.dirname/old_installed_caskfile.basename.sub(
+        old_token,
+        new_token,
+      )
```

Comment:
Are you sure we can't just do this here or is there some edge case I'm missing?

```suggestion
      new_installed_caskfile = old_installed_caskfile.sub(old_token, new_token)
```

Reply:
That's not the same thing since `sub` will only replace the first occurrence, which would be in `Caskroom/<token>/`, but we only want to replace it on the file name.

Reply:
They seem to print the same result on my machine when I tested them in the REPL.

```rb
~ $ brew irb
^[[A==> Interactive Homebrew Shell
Example commands available with: `brew irb --examples`
brew(main):001:0> cask = :virtualbox.c; nil
=> nil
brew(main):002:0> old_token = cask.token
=> "virtualbox"
brew(main):003:0> new_token = "virtualcats"
=> "virtualcats"
brew(main):004:0> old_installed_caskfile = cask.installed_caskfile.relative_path_from(cask.caskroom_path)
=> #<Pathname:.metadata/7.0.8,156879/20230423212244.154/Casks/virtualbox.rb>
brew(main):005:0> new_installed_caskfile = old_installed_caskfile.dirname/old_installed_caskfile.basename.sub(old_token, new_token)
=> #<Pathname:.metadata/7.0.8,156879/20230423212244.154/Casks/virtualcats.rb>
brew(main):006:0> new_installed_caskfile = old_installed_caskfile.sub(old_token, new_token) 
=> #<Pathname:.metadata/7.0.8,156879/20230423212244.154/Casks/virtualcats.rb>
```

Reply:
Right, it was needed since `old_installed_caskfile` was absolute before. Still, it makes more sense to me to only replace it in the basename.